transform_assumption.ml

```diff
@@ -58,22 +58,33 @@ module TargetSet = Set.Make(LabelTarget)
 
 (* Removes all empty checkpoints with no incoming references *)
 let remove_empty_osr : opt_prog = fun prog ->
-  let get_targets = function[@warning "-4"]
-    | Osr {target; cond; frame_maps} when cond <> [] ->
-      let add tgs {cont_pos} = TargetSet.union tgs (TargetSet.singleton cont_pos) in
-      List.fold_left add (TargetSet.singleton target) frame_maps
+  let get_targets func version seen = function[@warning "-4"]
+    | Osr {label; target; cond; frame_maps} ->
+      let my_pos = { func = func.name; version = version.label; pos = label } in
+      if cond <> [] || TargetSet.mem my_pos seen then begin
+        let add tgs {cont_pos} = TargetSet.union tgs (TargetSet.singleton cont_pos) in
```

Comment:
This is orthogonal to the current change but I would like to be sure that I understand correctly. The reason that we also add the continuation targets to the live target set is not that we know they will be `osr`-ed to; they may not be, as we only know that a `return` instruction will want to jump to there, but because we will use the label->pc translation only at runtime to materialize the continuations when taking this osr, and thus we need to preserve the label to do this translation.